### PR TITLE
Fix deployment tracking, add real integration test

### PR DESCRIPTION
**Pull request by stantonk**
Graphite upgrade broke deployment tracking, as they now expect a JSON array for the `tags`, which was causing HTTP 400 when posting to `statsd/events/`.

This PR goes a little beyond fixing this, because the code base as is was broken due to improper shell quoting that was never tested. `mendel deploy` fails as a result with this error:

```
======================================================================

mv: cannot move ‘524804b5afc7bf75986b626bfeaffa968c2bedd3’ to ‘'/srv/myservice/releases/20170725-225859-kevin-21dcc7bdaa7a1688823a55b7dfffecbdee5f7e8a'’: No such file or directory

========================================================================
```

notice how there are single `'` quotes around the destination directory. that doesn't actually exist on the remote server.

To avoid things like this in the future, I've added `tox` and a full-blown integration test that spins up a vagrant-like docker container that it actually deploys a dropwizard test service to and curls it to make sure it works.

### CODE REVIEW

File: mendel/core.py [hunk 1]
```diff
@@ -194,14 +194,11 @@ def _lpath(self, *args):
         return os.path.join(self._cwd, *args)
 
     def _rpath(self, *args):
-        return self.shell_quote(os.path.join(self._service_root, *args))
```
Comment (stantonk) [inline mendel/core.py]: this is where mendel was broken

File: mendel/core.py [hunk 2]
```diff
@@ -724,13 +721,14 @@ def _track_event_graphite(self, event):
         user = getpass.getuser()
         what = '%s %s %s on host %s' % (user, event, self._service_name, env.host_string)
         data = ''
-        tag_strs = (self._service_name, event)
-        tags = ' '.join(tuple(str(_) for _ in tag_strs))
+        tags = [str(s) for s in (self._service_name, event)]
```
Comment (stantonk) [inline mendel/core.py]: fix for deployment tracking in graphite

File: mendel/tests/test_core.py [hunk 3]
```diff
@@ -24,3 +26,16 @@ def test_tracking_does_not_break_deploy(self):
         mock_core._start_or_restart.assert_called_with()
         mock_core._track_event_graphite.assert_called_with('deployed')
         mock_core._track_event_api.assert_called_with('deployed')
+
+    @patch('urllib2.urlopen')
+    def test_track_event_graphite(self, urlopen):
```
Comment (stantonk) [inline mendel/tests/test_core.py]: regression unittest

File: mendel/tests/test_integration.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        curdir = os.path.dirname(os.path.abspath(__file__))
+        self.workingdir = os.path.join(curdir, '..', '..', 'examples', 'java', 'tgz')
+        fileloc = os.path.join(self.workingdir, MENDEL_TEST_FILE)
+        ssh_port = os.environ.get('STANTONK/FAT-CONTAINER_22_TCP')
+        with open(fileloc, 'wb') as f:
+            filecontents = MENDEL_TGZ_YAML % ssh_port
+            print filecontents
+            f.write(filecontents)
+
+    def test_tgz_mendel_deploy(self):
```
Comment (stantonk) [inline mendel/tests/test_integration.py]: integration test!

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Tox (https://tox.readthedocs.io/) is a tool for running tests
+# in multiple virtualenvs. This configuration file will run the
+# test suite on all supported python versions. To use it, "pip install tox"
+# and then run "tox" from this directory.
+
+[tox]
+envlist = py27
+
+[testenv]
+docker = stantonk/fat-container:1.0
```
Comment (stantonk) [inline tox.ini]: tox is amazing!
uses tox-docker to spin up docker container(s) before the tests run, and shuts them down after removal. handles ephemeral ports passing to the test code via environment variables that are dynamically generated. very cool.

this docker container is something i put together to emulate a vagrant box configured like how `sprout_java` works. it is a very iiiiiicky container that is horribly insecure, but works for this type of test.

https://github.com/stantonk/fat-container

File: mendel/tests/test_integration.py (hunk 4 above)
Comment (nicholassprout) [inline mendel/tests/test_integration.py]: Nice I'd love to migrate this: https://github.com/sproutsocial/sprout-hadoop/blob/master/hadoop-jobs/fabfile.py to to use the tgz version of mendel, but I was never able to get it to work.

Comment (stantonk) [inline mendel/tests/test_integration.py]: @nicholassprout ah, yea tgz is a little squirely to get right cuz it uses maven-assembly-plugin. i should add integration tests for the other bundle types (jar, deb, etc.) but at least on the mendel side of things they're all pretty similar (upload an archive file of some kind to /srv/myservice/releases, do the symlink swap, upstart restart)

**Pull request merged**